### PR TITLE
feat: implement settings

### PR DESCRIPTION
**Pull request by mad-it**

### CODE REVIEW

File: internal/eva/settings.go [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+const (
+	getSettingPath   = "/api/core/management/GetSetting"
+	setSettingPath   = "/api/core/management/SetSetting"
+	unsetSettingPath = "/api/core/management/UnsetSetting"
+)
+
+type SetSettingsRequest struct {
+	Key                string `json:"Key"`
+	Value              string `json:"Value"`
+	OrganizationUnitID int64  `json:"OrganizationUnitID,omitempty"`
```
Comment (williansouzagonc) [inline internal/eva/settings.go:21]: I believe in our case `OrganizationUnitID` will never be empty

File: internal/eva/settings.go [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+		SetBody(req).
+		Post(setSettingPath)
+
+	if err != nil {
+		tflog.Error(ctx, "An network error ocurred.", err)
+
+		return nil, err
+	}
+
+	if resp.StatusCode() != 200 {
```
Comment (williansouzagonc) [inline internal/eva/settings.go:38]: Does the EVA API always return status code `200` for every successful request? or could it also be `201` for creating a new resource or `204` for a deletion?

File: internal/eva/settings.go [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+		tflog.Info(ctx, "Request failed", "Status code", resp.StatusCode(), "body", resp.String())
+
+		return nil, errors.New(fmt.Sprintf("Request failed with error: %s", resp.String()))
```
Comment (williansouzagonc) [inline internal/eva/settings.go:41]: `resp.String()` converts the whole response to string or only the `body` ? 

I found this https://stackoverflow.com/questions/38673673/access-http-response-as-string-in-go, not sure if it gives same result

File: internal/eva/settings.go (hunk 1 above)
Comment (williansouzagonc) [reply internal/eva/settings.go:21]: Okay, its better to be compliant to their APIs

File: internal/provider/resource_eva_settings.go [hunk 4]
```diff
@@ -0,0 +1,175 @@
...
+	"github.com/hashicorp/terraform-plugin-log/tflog"
+
+	"github.com/mad-it/terraform-provider-eva/internal/eva"
+)
+
+type settingType struct{}
+
+func (t settingType) GetSchema(ctx context.Context) (tfsdk.Schema, diag.Diagnostics) {
+	return tfsdk.Schema{
+		MarkdownDescription: "Eva organization unit configration.",
```
Comment (williansouzagonc) [inline internal/provider/resource_eva_settings.go]:
```suggestion
		MarkdownDescription: "Eva organization unit settings configuration.",
```

File: internal/provider/resource_eva_settings.go [hunk 5]
```diff
@@ -0,0 +1,175 @@
...
+
+	// TODO: do we need an ID or some unique identifier for this resource?
+	_, err := r.provider.evaClient.SetSettings(ctx, eva.SetSettingsRequest{
+		Key:                data.Key.Value,
+		Value:              data.Value.Value,
+		OrganizationUnitID: data.OrganizationUnitID.Value,
+	})
+
+	if err != nil {
+		resp.Diagnostics.AddError("Creating setting unit failed.", fmt.Sprintf("Unable to create example, got error: %s", err))
```
Comment (williansouzagonc) [inline internal/provider/resource_eva_settings.go]:
```suggestion
		resp.Diagnostics.AddError("Creating setting unit failed.", fmt.Sprintf("Unable to create setting, got error: %s", err))
```

File: internal/provider/resource_eva_settings.go [hunk 6]
```diff
@@ -0,0 +1,175 @@
...
+		return
+	}
+
+	client_resp, err := r.provider.evaClient.GetSetting(ctx, eva.GetSettingRequest{
+		Key:                data.Key.Value,
+		OrganizationUnitID: data.OrganizationUnitID.Value,
+	})
+
+	if err != nil {
+		resp.Diagnostics.AddError("Creating setting unit failed.", fmt.Sprintf("Unable to create example, got error: %s", err))
```
Comment (williansouzagonc) [inline internal/provider/resource_eva_settings.go]:
```suggestion
		resp.Diagnostics.AddError("Getting setting data failed.", fmt.Sprintf("Unable to get setting, got error: %s", err))
```

File: internal/eva/settings.go (hunk 2 above)
Comment (mad-it) [reply internal/eva/settings.go:38]: Not full REST complaint, so sadly enough that is not the case.

File: internal/provider/resource_eva_settings.go [hunk 7]
```diff
@@ -0,0 +1,175 @@
...
+		return
+	}
+
+	if strings.HasPrefix(client_resp.Value, "********") &&
```
Comment (williansouzagonc) [inline internal/provider/resource_eva_settings.go]: we could do something like (forgive my go code), an empty if doens't look correct to me

``` 
var isSensitiveData  boolean 

isSensitiveData = strings.HasPrefix(client_resp.Value, "********") && strings.HasSuffix(client_resp.Value, data.Value.Value[len(data.Value.Value)-4:])

if isSensitiveData === false {
  data.Value = types.String{Value: client_resp.Value}
}
```

File: internal/eva/settings.go (hunk 3 above)
Comment (mad-it) [reply internal/eva/settings.go:41]: `resp.String()` is indeed the body of the response. Not the entire response

File: internal/provider/resource_eva_settings.go [hunk 8]
```diff
@@ -0,0 +1,175 @@
...
+
+	// TODO: do we need an ID or some unique identifier for this resource?
+	_, err := r.provider.evaClient.SetSettings(ctx, eva.SetSettingsRequest{
+		Key:                data.Key.Value,
+		Value:              data.Value.Value,
+		OrganizationUnitID: data.OrganizationUnitID.Value,
+	})
+
+	if err != nil {
+		resp.Diagnostics.AddError("Updating setting unit failed.", fmt.Sprintf("Unable to update OU, got error: %s", err))
```
Comment (williansouzagonc) [inline internal/provider/resource_eva_settings.go]:
```suggestion
		resp.Diagnostics.AddError("Updating setting unit failed.", fmt.Sprintf("Unable to update setting, got error: %s", err))
```

File: internal/provider/resource_eva_settings.go [hunk 9]
```diff
@@ -0,0 +1,175 @@
...
+		return
+	}
+
+	_, err := r.provider.evaClient.UnsetSettings(ctx, eva.UnsetSettingsRequest{
+		Key:                data.Key.Value,
+		OrganizationUnitID: data.OrganizationUnitID.Value,
+	})
+
+	if err != nil {
+		resp.Diagnostics.AddError("Deleting setting unit failed.", fmt.Sprintf("Unable to delete OU, got error: %s", err))
```
Comment (williansouzagonc) [inline internal/provider/resource_eva_settings.go]:
```suggestion
		resp.Diagnostics.AddError("Deleting setting unit failed.", fmt.Sprintf("Unable to delete setting, got error: %s", err))
```

**Review by williansouzagonc — APPROVED** [review]
Great job man 🚀  ! I don't have much of a GO knowledge so not sure about best practises but code looks pretty easy to read and understand ! I left few comments, feel free to address them

**Pull request merged**